File: apps/ios/ITTRehber/Models/Listing.swift
import Foundation

struct Listing: Codable, Identifiable, Hashable {
    let id: UUID
    let name: String
    let directories: [String]
    let kantons: [String]
    let category: String?
    let subCategory: String?
    let address: String?
    let phone: String?
    let email: String?
    let website: String?
    let description: String?
    let imageURL: String?
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case directories
        case kantons
        case category
        case subCategory = "sub_category"
        case address
        case phone
        case email
        case website
        case description
        case imageURL = "image_url"
        case updatedAt = "updated_at"
    }
}

struct ListingPage: Codable {
    let items: [Listing]
    let total: Int
    let page: Int
    let pageSize: Int

    enum CodingKeys: String, CodingKey {
        case items
        case total
        case page
        case pageSize = "page_size"
    }
}

struct ListingSubmitInput: Codable {
    var name: String
    var contactPerson: String?
    var directories: [String]
    var kantons: [String]
    var category: String?
    var address: String?
    var phone: String?
    var phonePublic: Bool
    var email: String?
    var emailPublic: Bool
    var website: String?
    var description: String?
    var imageURL: String?
    var package: ListingPackage?

    enum CodingKeys: String, CodingKey {
        case name
        case contactPerson = "contact_person"
        case directories
        case kantons
        case category
        case address
        case phone
        case phonePublic = "phone_public"
        case email
        case emailPublic = "email_public"
        case website
        case description
        case imageURL = "image_url"
        case package
    }
}

enum ListingPackage: String, Codable, CaseIterable, Identifiable {
    case months_3
    case months_6
    case months_12

    var id: String { rawValue }

    var titleTR: String {
        switch self {
        case .months_3: return "3 ay — 60 CHF"
        case .months_6: return "6 ay — 100 CHF"
        case .months_12: return "12 ay — 180 CHF"
        }
    }

    var amountCHF: Int {
        switch self {
        case .months_3: return 60
        case .months_6: return 100
        case .months_12: return 180
        }
    }
}

struct InvoiceSummary: Codable, Identifiable {
    let id: UUID
    let listingId: UUID
    let invoiceNumber: String
    let amountChf: Int  // cents
    let package: String
    let issuedAt: Date
    let dueAt: Date?
    let paidAt: Date?
    let pdfURL: String?

    enum CodingKeys: String, CodingKey {
        case id
        case listingId = "listing_id"
        case invoiceNumber = "invoice_number"
        case amountChf = "amount_chf"
        case package
        case issuedAt = "issued_at"
        case dueAt = "due_at"
        case paidAt = "paid_at"
        case pdfURL = "pdf_url"
    }
}

File: apps/backend/app/models/listing.py
"""Listing — directory entry. Implements the PRD §6.3 state machine."""

from __future__ import annotations

import enum
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text, func
from sqlalchemy.dialects.postgresql import ARRAY, ENUM, UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db import Base


class ListingStatus(str, enum.Enum):
    pending = "pending"
    active = "active"
    rejected = "rejected"
    suspended = "suspended"
    expired = "expired"
    archived = "archived"


class ListingPackage(str, enum.Enum):
    months_3 = "months_3"
    months_6 = "months_6"
    months_12 = "months_12"


# Allowed transitions per PRD §6.3.
ALLOWED_TRANSITIONS: dict[ListingStatus, set[ListingStatus]] = {
    ListingStatus.pending: {ListingStatus.active, ListingStatus.rejected},
    ListingStatus.active: {ListingStatus.suspended, ListingStatus.expired, ListingStatus.pending},
    ListingStatus.rejected: set(),
    ListingStatus.suspended: {ListingStatus.active, ListingStatus.archived},
    ListingStatus.expired: {ListingStatus.active, ListingStatus.archived},
    ListingStatus.archived: set(),
}


class Listing(Base):
    __tablename__ = "listings"

    id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), primary_key=True, default=uuid4)

    name: Mapped[str] = mapped_column(String(200), nullable=False)
    contact_person: Mapped[str | None] = mapped_column(String(120))

    # Multi-value: a listing can appear in multiple directories and span multiple kantons
    # (PRD §5.7: single fee covers multiple directories and kantons).
    directories: Mapped[list[str]] = mapped_column(ARRAY(String(32)), nullable=False, default=list)
    kantons: Mapped[list[str]] = mapped_column(ARRAY(String(2)), nullable=False, default=list)

    category: Mapped[str | None] = mapped_column(String(64))
    sub_category: Mapped[str | None] = mapped_column(String(64))

    address: Mapped[str | None] = mapped_column(String(256))
    phone: Mapped[str | None] = mapped_column(String(64))
    phone_public: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    email: Mapped[str | None] = mapped_column(String(320))
    email_public: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    website: Mapped[str | None] = mapped_column(String(512))

    description: Mapped[str | None] = mapped_column(Text)
    image_url: Mapped[str | None] = mapped_column(String(1024))

    owner_id: Mapped[UUID | None] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), index=True
    )

    status: Mapped[ListingStatus] = mapped_column(
        ENUM(ListingStatus, name="listing_status"),
        default=ListingStatus.pending,
        nullable=False,
        index=True,
    )

    package: Mapped[ListingPackage | None] = mapped_column(
        ENUM(ListingPackage, name="listing_package")
    )
    paid_until: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    rejection_reason: Mapped[str | None] = mapped_column(String(64))
    rejection_notes: Mapped[str | None] = mapped_column(Text)

    approved_by: Mapped[UUID | None] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL")
    )
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    # ---- State machine -------------------------------------------------

    def can_transition_to(self, target: ListingStatus) -> bool:
        return target in ALLOWED_TRANSITIONS.get(self.status, set())

    def transition_to(self, target: ListingStatus) -> None:
        if not self.can_transition_to(target):
            raise ValueError(
                f"Illegal transition {self.status.value} -> {target.value}"
            )
        self.status = target
